synapse/replication/http/account_data.py

```diff
@@ -76,16 +76,21 @@ class ReplicationRemoveUserAccountDataRestServlet(ReplicationEndpoint):
 
         POST /_synapse/replication/remove_user_account_data/:user_id/:type
 
-        {
-            "content": { ... },
-        }
+        {}
 
     """
 
     NAME = "remove_user_account_data"
     PATH_ARGS = ("user_id", "account_data_type")
     CACHE = False
 
+    # In the case of attempting to remove account data that doesn't exist,
+    # the account data stream will not be incremented by the handler of this
+    # replication endpoint. Therefore, we don't wait for the account data
+    # stream to update after issuing this replication request. Even with this
+    # off, callers will still wait for a response to the request.
+    WAIT_FOR_STREAMS = False
```

Comment:
Does the client get stuck waiting if WAIT_FOR_STREAMS is turned on? I'd expect it to handle the case where no streams have been updated just fine.

Reply:
Indeed, because by definition it waits for the account data stream to be incremented. In the case of issuing a deletion of an account data object that doesn't exist, the stream is not modified.

Thus the client gets stuck waiting for something that will never come.

Reply:
I'm very confused about the waiting behaviour now.

On the responding side, the worker attaches a dict containing the current position of ~~every stream~~ the cache invalidation and account data streams, regardless of whether they have been written to.
https://github.com/matrix-org/synapse/blob/c78c67c5a909c6749f25b251d46be3df8f56f8c2/synapse/replication/http/_base.py#L446-L450

On the requesting side, we do a wait for every stream in the response, passing the responding worker's stream position:
https://github.com/matrix-org/synapse/blob/c78c67c5a909c6749f25b251d46be3df8f56f8c2/synapse/replication/http/_base.py#L347-L355

but I would expect `wait_for_stream_position` to be a no-op when given a stream position in the past... unless it's horribly broken?

Reply:
Hmm, good point. i think the problem may actually be that we're incrementing the `account_data` stream ID even if we don't write to the database, as `self._account_data_id_gen.get_next()` will increment `self._current`.

https://github.com/matrix-org/synapse/blob/d0c713cc85f094c323b2ba3f02d8ac411a7f0705/synapse/storage/databases/main/account_data.py#L805-L813

https://github.com/matrix-org/synapse/blob/65d03866936adb144631d263a8539a2cb060fd43/synapse/storage/util/id_generators.py#L222-L239

So we would somehow need to decrement the stream if we didn't write to the database. Though, is that even possible with multiple writers?

Reply:
That sounds somewhat complicated / error prone! I think we should strive to be making our APIs simpler and less foot-gunny.

Reply:
Unfortunately to achieve this, I think we may need to do an extra query against the database. Currently we advance the stream ID position and then attempt to perform an `UPDATE` against the `account_data` table:

https://github.com/matrix-org/synapse/blob/d0c713cc85f094c323b2ba3f02d8ac411a7f0705/synapse/storage/databases/main/account_data.py#L757-L770

If the `UPDATE` comes back as having made no changes, then ideally we'd then decrement the stream ID position again. I don't know how tenable this is however.

Alternatively we separate the check for whether this would make a change into an initial `SELECT`, and then if we're convinced that the action would not be a no-op, _then_ we advance the stream position.

But anyhow, I think the ideal scenario would be to:

1. Lock the stream ID generator
2. Get the next stream ID
3. Attempt to UPDATE against the database
4. If a change was actually made, then increment the stream ID position
5. If not, then do not modify the stream ID position

But the stream ID generators don't currently expose any public interface for doing the above afaict.

Reply:
Considering https://github.com/matrix-org/synapse/pull/14820 was intended to help fix race conditions, I'm minded to just take the extra query rather than expose ourselves to race condition bugs.

Reply:
Perhaps the root of the issue was the bug that @erikjohnston fixed in #15150?

Reply:
> Perhaps the root of the issue was the bug that @erikjohnston fixed in #15150

Oops, our comments clashed.

Alas, switching to `develop`, cherry-picking the first fix from this PR (972ebd13f0854106dfaf37221c56cce7facbd523) and re-running the `TestLogout` Complement tests with `WORKER_TYPES=account_data` now passes, whereas presumably before https://github.com/matrix-org/synapse/pull/15150 it would not.

Is the extra work in this PR still relevant then? I think not, as it introduces an additional DB query.

The `get_next_txn` method added in ae14dfcb4fcecd26f49a0e8951359c60dd5a3018 is still useful, but that has already been split out to a new PR.

`remove_account_data_for_room` and `remove_account_data_for_user` having their return types changed from `Optional[int]` to `int` is also useful, so I'll keep that as well.

The rest of this PR I'll strip out, which should significantly reduce the line diff :)